engine/core/src/main/java/org/datacleaner/reference/regexswap/RegexSwapClient.java

```diff
@@ -19,150 +19,101 @@
-    }
-
-    public List<Regex> getRegexes(final Category category) {
-        final List<Regex> regexes = new ArrayList<>();
-        final Node rootNode = HttpXmlUtils.getRootNode(_httpClient, category.getDetailsUrl());
-        final List<Node> regexNodes = HttpXmlUtils.getChildNodesByName(rootNode, "regex");
-        for (final Node regexNode : regexNodes) {
-
-            final String name = HttpXmlUtils.getChildNodeText(regexNode, "name");
-            final String description = HttpXmlUtils.getChildNodeText(regexNode, "description");
-            final String expression = HttpXmlUtils.getChildNodeText(regexNode, "expression");
-            final String author = HttpXmlUtils.getChildNodeText(regexNode, "author");
-            final long timestamp = Long.parseLong(HttpXmlUtils.getChildNodeText(regexNode, "timestamp"));
-            final int positiveVotes = Integer.parseInt(HttpXmlUtils.getChildNodeText(regexNode, "positiveVotes"));
-            final int negativeVotes = Integer.parseInt(HttpXmlUtils.getChildNodeText(regexNode, "negativeVotes"));
-            final String detailsUrl = HttpXmlUtils.getChildNodeText(regexNode, "detailsUrl");
-
-            final List<Category> categories;
-            Regex regex = _regexes.get(name);
-            if (regex == null) {
-                categories = new ArrayList<>();
-                regex = new Regex(name, description, expression, author, timestamp, positiveVotes, negativeVotes,
-                        detailsUrl, categories);
-            } else {
-                categories = regex.getCategories();
-                if (!categories.contains(category)) {
-                    categories.add(category);
-                }
-            }
-
-            _regexes.put(name, regex);
-
-            regexes.add(regex);
-        }
-        return regexes;
-    }
+	private static final URI REGEXES_URI = URI.create("https://datacleaner.github.io/content/regexes.json");
+
+	private final Map<String, Category> _categories = new HashMap<>();
+	private final Map<String, Regex> _regexes = new HashMap<>();
+	private final HttpClient _httpClient;
+
+	public RegexSwapClient(final HttpClient httpClient) {
+		_httpClient = httpClient;
+	}
+
+	public Category getCategoryByName(final String name) {
+		if (_regexes.isEmpty()) {
+			refreshRegexes();
+		}
+		return _categories.get(name);
+	}
+
+	public Regex getRegexByName(final String name) {
+		if (_regexes.isEmpty()) {
+			refreshRegexes();
+		}
+		return _regexes.get(name);
+	}
+
+	@SuppressWarnings("unchecked")
+	public void refreshRegexes() {
+		final HttpUriRequest request = RequestBuilder.get(REGEXES_URI).build();
+		final Map<String, ?> root;
+		try {
+			final HttpResponse response = _httpClient.execute(request);
+			final String json = EntityUtils.toString(response.getEntity());
+			final ObjectMapper objectMapper = new ObjectMapper();
+			root = objectMapper.readValue(json, Map.class);
+		} catch (IOException e) {
+			throw new UncheckedIOException(e);
```

Comment:
Shouldn't this warn the user, or is it handled somewhere else?

Reply:
I didn't really mean "more specific". I meant more like a proper "we couldn't contact the server" local warning instead of just dumping it up the chain.

But I might misremember how we handle exceptions and this is all handled properly somewhere (though "someone will probably handle it" seems like a bad strategy).

Reply:
I see what you mean, but since this is a engine/backend component I think that throwing an exception is most appropriate. We could add more logging and/or make a specific exception for this scenario. But then again I do kinda think this is also just about wrapping the checked exception in an unchecked exception and deciding not to try and handle it, because we have no good way to recover from it here.